src/merkle_tree.cpp

```diff
@@ -119,18 +119,23 @@ namespace {
 		, std::vector<bool> const& verified)
 	{
 		INVARIANT_CHECK;
-		TORRENT_ASSERT(mask.size() == size());
-		if (size() != mask.size()) return;
+
+		// The size of the mask should not exceed the size of the tree, but a mask larger than
+		// the tree can be encountered in the "resume data" for some reason, perhaps there is
+		// a bug in the "resume data" generation algorithm.
+		// So we just process mask items count up to tree size.
+		int const mask_size = int(std::min(mask.size(), size()));
 
 		int const first_block = block_layer_start();
 		int const end_block = first_block + m_num_blocks;
 
-		TORRENT_ASSERT(first_block < int(mask.size()));
-		TORRENT_ASSERT(end_block <= int(mask.size()));
+		TORRENT_ASSERT(first_block < int(size()));
+		TORRENT_ASSERT(end_block <= int(size()));
 
 		// if the mask covers all blocks, go straight to block_layer
 		// mode, and validate
-		if (std::all_of(mask.begin() + first_block, mask.begin() + end_block, identity()))
+		if ((first_block < mask_size) && (end_block <= mask_size)
```

Comment:
```suggestion
		if (end_block <= mask_size
```
We're guaranteed that `first_block` < `end_block`, right?

Reply:
> We're guaranteed that `first_block` < `end_block`, right?

At least we assert on this indirectly:
```c++
TORRENT_ASSERT(first_block < int(size()));
TORRENT_ASSERT(end_block <= int(size()));
```